module/zfs/dmu.c

```diff
@@ -766,6 +766,50 @@ dmu_prefetch(objset_t *os, uint64_t object, int64_t level, uint64_t offset,
 	dnode_rele(dn, FTAG);
 }
 
+/* XXX comment me -- robn, 2024-02-07 */
+int
+dmu_prefetch_wait(objset_t *os, uint64_t object, uint64_t offset, uint64_t size,
+    uint32_t flags)
+{
+	dnode_t *dn;
+	dmu_buf_t **dbp;
+	int numbufs, err = 0;
+
+	err = dnode_hold(os, object, FTAG, &dn);
+	if (err != 0)
+		return (err);
+
+	while (size > 0) {
+		uint64_t mylen = MIN(size, dmu_prefetch_max);
+
+		/*
+		 * NB: we could do this block-at-a-time, but it's nice
+		 * to be reading in parallel.
+		 *
+		 * XXX -- can we use dbuf_prefetch_impl() to avoid filling
+		 * the dbuf cache with decompressed ddt data which we don't
+		 * need right now?
+		 */
+		err = dmu_buf_hold_array_by_dnode(dn, offset, mylen,
+		    TRUE, FTAG, &numbufs, &dbp, flags);
+		if (err)
+			break;
+
+		dmu_buf_rele_array(dbp, numbufs, FTAG);
```

Comment:
We definitely don't need decompression here, that may be too expensive for hundreds of gigabytes and completely wipe dbuf cache.  I think much cleaner would be to add to dmu_prefetch_impl() a callback argument similar to dbuf_prefetch_impl() and returning number of prefetches issued, make the callback function count completions, and wait till all requests get back.

Reply:
Yep, it should not need to decompress.  Fixed to only read into ARC and never inflate into dbuf cache